src/base/memory_monitor_sr.cpp

```diff
@@ -0,0 +1,235 @@
+/*
+ *
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * memory_monitor_sr.cpp - implementation of memory monitoring manager
+ */
+
+#include <stdexcept>
+#include <cstring>
+
+#include "perf.hpp"
+#include "error_manager.h"
+#include "memory_monitor_sr.hpp"
+
+#include "log_impl.h"
+
+namespace cubperf
+{
+  memory_monitor *mmon_Gl = nullptr;
+
+  const char *mmon_subcomponent::get_name ()
+  {
+    return m_subcomp_name.c_str ();
+  }
+
+  void mmon_subcomponent::add_cur_stat (uint64_t size)
+  {
+    m_cur_stat.fetch_add (size);
+  }
+
+  void mmon_subcomponent::sub_cur_stat (uint64_t size)
+  {
+    m_cur_stat.fetch_sub (size);
+  }
+
+  const char *mmon_component::get_name ()
+  {
+    return m_comp_name.c_str ();
+  }
+
+  void mmon_component::add_stat (uint64_t size, int subcomp_idx, bool init, bool expand)
+  {
+    /* description of add_stat(..).
+     * 1) if init == true, add init_stat
+     * 2) then, add cur_stat
+     * 3) compare with peak_stat
+     * 4) if cur_stat(new) > peak_stat, update peak_stat
+     * 5) if expand == true, expand_count++
+     * 6) call subcomponent->add_cur_stat(size) */
+    return;
+  }
+
+  void mmon_component::sub_stat (uint64_t size, int subcomp_idx, bool init)
+  {
+    /* description of sub_stat(size, init).
+     * 1) if init == true, sub init_stat
+     * 2) then, sub cur_stat
+     * 3) call component->sub_stat(size, subcomp_idx, init) */
+    return;
+  }
+
+  int mmon_component::is_subcomp_exist (const char *subcomp_name)
+  {
+    for (int i = 0; i < m_subcomponent.size (); i++)
+      {
+	if (!strcmp (subcomp_name, m_subcomponent[i]->get_name()))
+	  {
+	    return i;
+	  }
+      }
+
+    /* if not exist, return mmon_module::MAX_COMP_IDX */
+    return mmon_module::MAX_COMP_IDX;
+  }
+
+  int mmon_component::add_subcomponent (const char *name)
+  {
+    return 0;
+  }
+
+  mmon_module::mmon_module (const char *module_name, const MMON_COMP_INFO *info)
+    : m_module_name (module_name)
+  {
+    /* register component and subcomponent information
+     * add component and subcomponent */
+    int cnt = 0;
+    while (info[cnt].id != MMON_STAT_LAST)
+      {
+	bool comp_skip = false;
+	bool subcomp_skip = false;
+	int comp_idx = mmon_module::MAX_COMP_IDX, subcomp_idx = mmon_module::MAX_COMP_IDX;
```

Comment:
1. Why is `MAX_COMP_IDX` needed? Can't we use just -1 for the NULL comp or subcomp case?
2. What happens if they end up having `mmon_module::MAX_COMP_IDX`. Is it allowed?

Reply:
MAX_COMP_IDX has the value of 0x0000FFFF. 
comp_idx and subcomp_idx will be combined into a single 32-bit value and stored in a vector. Each of them cannot have a value exceeding 16 bits. Therefore, MAX_COMP_IDX represents an unreachable (should not be reached) value. It can be checked during the implementation of the function. If the value of a specific comp_idx or subcomp_idx is equal to MAX_COMP/SUBCOMP_IDX, it is considered that there is no component or subcomponent mapped to that STAT_ID.